### PR TITLE
Implement static import for `ISequentialStream` (#474)

### LINKED ISSUE #474: Returning `out` arguments breaks when they must be preallocated
The standard treatment of argument directions is as follows:

- An `'in'` argument must be provided and is not returned.
- An `'out'` argument is returned by the function and must not be provided.
- An `'in', 'out'` argument must be provided and is returned.

This model breaks down when a parameter marked as `'out'` must be preallocated with some defined amount of memory. This happens for example in [ISequentialStream::Read](https://learn.microsoft.com/en-us/windows/win32/api/objidl/nf-objidl-isequentialstream-read) (or `RemoteRead` in my case, which has the same signature). The parameters are given by
```python
        (['out'], POINTER(c_ubyte), 'pv'),
        (['in'], c_ulong, 'cb'),
        (['out'], POINTER(c_ulong), 'pcbRead')
```
This call only works if `pv` is pre-allocated with `cb` bytes, which cannot be done in the current model.

I resorted to calling the base method `ISequentialStream._ISequentialStream__com_RemoteRead` with three arguments. My question is whether I am overlooking a more elegant solution. If that is not the case, I was wondering if there is a better way to treat this problem.

One other option is to change the generated header and set the first parameter to be `['in', 'out']`, but this breaks if one re-generates the file and is also technically incorrect.

### PR DESCRIPTION
**Pull request by jonschz**
See #474. I migrated the code from the old `istream-issue` branch to the latest `main` branch. Let me know if anything else needs to be done.

Closes #474

### CODE REVIEW

File: comtypes/objidl.py
Comment (junkmd) [inline comtypes/objidl.py:1]: I feel that this module should be renamed to `stream.py`.

I originally named it `objidl.py` based on `objidl.h`.
However, I realized that if we continue to maintain it in this approach, we will have to create and define utilities related to `IStream` in `combaseapi.py` and `shlwapi.py`, which will result in a loss of cohesion.

What do you think?

**Review by junkmd — CHANGES_REQUESTED** [review]
I have added the parts of the test that I was concerned about.

There are no problems with the production codebase or its behavior.

As remaining tasks to merge this PR, we need to work on naming and writing comments.

**Comment by junkmd** [comment]
I have come across a repository that makes `RemoteRead` usable by directly modifying the module generated in `comtypes.gen` (such as changing `pv` to `["in", "out"]`). If this PR is merged, that approach will be no more, and that codebase may not work.

However, that approach involves modifying the codebase within the `comtypes` package. In other words, it is essentially synonymous with forking and applying a custom patch for use.

Therefore, I think that those codebases becoming inoperable due to this change does not constitute "breaking backward compatibility".

File: comtypes/objidl.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    def RemoteRead(self, cb: int) -> Tuple["Array[c_ubyte]", int]:
+        """Reads a specified number of bytes from the stream object into memory
+        starting at the current seek pointer.
+        """
+        # Behaves as if `pv` is pre-allocated with `cb` bytes by the high level func.
+        pv = (c_ubyte * cb)()
+        pcb_read = pointer(c_ulong(0))
+        self.__com_RemoteRead(pv, c_ulong(cb), pcb_read)  # type: ignore
+        return pv, pcb_read.contents.value
```
[the remark this answers is not in the document]
Comment (jonschz) [reply comtypes/objidl.py]: I added a small comment.

File: comtypes/objidl.py
Comment (jonschz) [reply comtypes/objidl.py:1]: Makes sense, I changed it.

File: comtypes/objidl.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+
+class ISequentialStream(IUnknown):
+    """Defines methods for the stream objects in sequence."""
+
+    _iid_ = GUID("{0C733A30-2A1C-11CE-ADE5-00AA0044773D}")
+    _idlflags_ = []
+
+    _methods_ = [
+        COMMETHOD(
```
[the remark this answers is not in the document]
Comment (jonschz) [reply comtypes/objidl.py]: I added an explanation. Feel free to change it if you feel that it can be improved. Otherwise, feel free to resolve.

**Comment by jonschz** [comment]
> In other words, it is essentially synonymous with forking and applying a custom patch for use.

I agree with that. If all the downstream code we break is code that changed this packages behaviour, that is not a "breaking change" in my opinion, since they did not use part of our official API. Giving them a heads up won't hurt, but whenever you do something like that, you pretty much have to do version pinning (or blame yourself if an external update breaks your package).

The part of the official API that we *did* change has never worked in the past to begin with, so it's quite a stretch to call that a "breaking change" in my opinion.

**Comment by junkmd** [comment]
> Giving them a heads up won't hurt, but whenever you do something like that, you pretty much have to do version pinning (or blame yourself if an external update breaks your package).

I am planning to include this change in the release of version 1.4.5.
This means that we will explicitly state that the 'custom patch' approach to `RemoteRead` under `comtypes.gen` can only be used up to version 1.4.4.

For example, if there is a 'custom patch' version of `RemoteRead` that 
> sets the first parameter to be `['in', 'out']`

as you mentioned in https://github.com/enthought/comtypes/issues/474#issue-1536002087, it is necessary to make the following changes:
```diff
buffer_size = 1024
- pv = (c_ubyte * buffer_size)()
- read_buffer, data_read = stream.RemoteRead(pv, buffer_size)
+ read_buffer, data_read = stream.RemoteRead(buffer_size)
```

Since it is impossible for us to investigate how widely the 'custom patch' is used in the world, I do NOT plan to find and go around individual repositories to encourage caution.

> The part of the official API that we _did_ change has never worked in the past to begin with, so it's quite a stretch to call that a "breaking change" in my opinion.

I agree with this.
I think making something that was causing some kind of error at the time of the call usable is in the realm of 'bug fixes'.
I do NOT think it's a "breaking change" if a user's ad hoc workaround starts causing errors when a bug has been fixed upstream.

I will also mention that this change only affects `ISequentialStream` defined within the module generated by `GetModule` in `comtypes.gen`, and does not break codebases that define and implement `ISequentialStream` (and `tagSTATSTG` and `IStream`) statically on their own.

**Comment by junkmd** [comment]
I recognize that there are (other than `ISequentialStream`'s `RemoteRead`) methods that must be overriden because the codebase generated by `codegenerator` is not usable as is.
However, in this project, I am reluctant to statically define those interfaces and methods anything and everything.
Further discussion is needed in a separate scope from #474 for interfaces other than `ISequentialStream`.

I am considering accepting #578 is because `IStream` is the de facto standard for handling streams in COM, and I think it is in the community's profit to make `RemoteRead` usable.

**Review by junkmd — APPROVED** [review]

**Comment by jonschz** [comment]
> I recognize that there are (other than `ISequentialStream`'s `RemoteRead`) methods that must be overriden because the codebase generated by `codegenerator` is not usable as is.

I am sure that other examples will show up over time. As I don't see an automatic fix at the moment, I suggest we just implement them by hand as they come in. The fact that `ISequentialStream` is (afaik) the only one reported so far suggests that the problem is not extremely frequent.

Thank you for your efforts and contributions to this project!

**Pull request merged**